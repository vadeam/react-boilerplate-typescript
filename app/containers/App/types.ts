import { Record } from 'immutable'

import { Repo } from 'containers/RepoListItem/types'

export interface AppActionTypes {
  LOAD_REPOS: string
  REPOS_LOADED: string
  REPO_LOADING_ERROR: string
}

export interface LoadReposAction extends Action<AppActionTypes> {}

export interface LoadReposSuccessAction extends Action<AppActionTypes> {
  readonly repos: Nullable<Repo[]>
  readonly username: string
}

export interface LoadReposErrorAction<T = object> extends Action<AppActionTypes> {
  readonly error: T
}

export type AppActions = LoadReposAction | LoadReposSuccessAction | LoadReposErrorAction

export interface AppActionCreators {
  loadRepos: () => LoadReposAction
  reposLoaded: (repos: Nullable<Repo[]>, username: string) => LoadReposSuccessAction
  repoLoadingError: (error: object) => LoadReposErrorAction
}

export interface UserData {
  readonly repos: Nullable<Repo[]>
}

export const UserDataRecord = Record<UserData>({ repos: null }, 'UserDataRecord')

export interface AppState extends InheritedReducerState {
  readonly currentUser: string
  readonly userData: UserData
}

export const AppStateRecord = Record<AppState>(
  {
    isFetching: false,
    currentUser: '',
    userData: UserDataRecord(),
    error: null,
  },
  'AppStateRecord',
)
