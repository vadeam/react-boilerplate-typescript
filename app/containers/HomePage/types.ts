import { Record } from 'immutable'

export interface HomeActionTypes {
  CHANGE_USERNAME: string
}

export interface HomeAction extends Action<HomeActionTypes> {
  readonly username: string
}

export interface HomeActionCreators {
  changeUsername: (username: string) => HomeAction
}

export interface HomeState extends InheritedReducerState {
  readonly username: string
}

export const HomeStateRecord = Record<HomeState>(
  {
    username: '',
  },
  'HomeStateRecord',
)
