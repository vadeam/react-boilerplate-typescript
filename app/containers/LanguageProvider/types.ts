import { Record } from 'immutable'

import { LOCALE_EN, LOCALE_DE, DEFAULT_LOCALE } from 'constants/locales'

declare global {
  type Locales = typeof LOCALE_EN | typeof LOCALE_DE
}

export type LocalMessages = { [p: string]: string }

export interface LanguageActionTypes {
  CHANGE_LOCALE: string
  SET_LOCALE_MESSAGES: string
}

export interface ChangeLocaleAction extends Action<LanguageActionTypes> {
  readonly locale: Locales
}

export interface SetLocaleMessagesAction extends Action<LanguageActionTypes> {
  readonly messages: LocalMessages
}

export interface LanguageActionCreators {
  changeLocale: (locale: Locales) => ChangeLocaleAction
  setLocaleMessages: (messages: LocalMessages) => SetLocaleMessagesAction
}

export interface LanguageProviderState extends InheritedReducerState {
  readonly locale: Locales
  readonly messages: LocalMessages
}

export const LanguageProviderStateRecord = Record<LanguageProviderState>(
  {
    locale: DEFAULT_LOCALE,
    messages: {},
  },
  'LanguageProviderStateRecord',
)
