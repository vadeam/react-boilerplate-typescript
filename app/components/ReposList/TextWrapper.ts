import styled from 'styles/styled-components'

const TextWrapper = styled.div`
  line-height: 1.5em;
  color: lightgray;
  text-align: center;
  padding: 0 1em;
`

export default TextWrapper
