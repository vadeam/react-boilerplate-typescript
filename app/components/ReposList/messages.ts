import { defineMessages } from 'react-intl'

export default defineMessages({
  userHint: {
    id: 'containers.HomePage.user.hint',
    defaultMessage: 'Type user name in the input above and press Enter',
  },
})
