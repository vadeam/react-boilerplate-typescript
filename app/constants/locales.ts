export const LOCALE_DE = 'de'
export const LOCALE_EN = 'en'
export const DEFAULT_LOCALE = LOCALE_EN

export const appLocales: Locales[] = [LOCALE_EN, LOCALE_DE]
export const CUSTOM_LOCALE_STORAGE_KEY = 'custom-locale'
