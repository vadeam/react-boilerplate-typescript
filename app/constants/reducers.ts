export const REDUCER_KEY_APP = 'app'
export const REDUCER_KEY_LANGUAGE = 'language'
export const REDUCER_KEY_ROUTER = 'router'
export const REDUCER_KEY_HOME = 'home'
// for testing purposes
export const REDUCER_KEY_TEST = 'test'
