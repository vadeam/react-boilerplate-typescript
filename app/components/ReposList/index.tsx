import * as React from 'react'
import { useIntl } from 'react-intl'

import List from 'components/List'
import Wrapper from 'components/List/Wrapper'
import ListItem from 'components/ListItem'
import LoadingIndicator from 'components/LoadingIndicator'
import RepoListItem from 'containers/RepoListItem'

import { Repo } from '../../containers/RepoListItem/types'

import TextWrapper from './TextWrapper'
import messages from './messages'

type ReposListProps = {
  isFetching?: boolean
  error?: Nullable<object | boolean>
  repos?: Nullable<Repo[]>
}

function ReposList({ isFetching, error, repos }: ReposListProps) {
  const { formatMessage } = useIntl()

  if (isFetching) {
    return <List component={LoadingIndicator} />
  }

  if (error) {
    return <List component={() => <ListItem item="Something went wrong, please try again!" />} />
  }

  if (repos) {
    return <List<Repo> items={repos} component={RepoListItem} />
  }

  return (
    <Wrapper>
      <TextWrapper>{formatMessage({ ...messages.userHint })}</TextWrapper>
    </Wrapper>
  )
}

export default ReposList
