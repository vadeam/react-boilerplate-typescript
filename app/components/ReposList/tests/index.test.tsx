import { render } from '@testing-library/react'
import React from 'react'
import { IntlProvider } from 'react-intl'
import { Provider } from 'react-redux'

import { REDUCER_KEY_APP } from '../../../constants/reducers'
import configureStore from '../../../configureStore'
import { Repo } from '../../../containers/RepoListItem/types'
import history from '../../../utils/history'
import ReposList from '../index'
import { DEFAULT_LOCALE } from '../../../constants/locales'

describe('<ReposList />', () => {
  it('should render the loading indicator when its loading', () => {
    const { container } = render(
      <IntlProvider locale={DEFAULT_LOCALE}>
        <ReposList isFetching />
      </IntlProvider>,
    )
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render an error if loading failed', () => {
    const { queryByText } = render(
      <IntlProvider locale={DEFAULT_LOCALE}>
        <ReposList isFetching={false} error={{ message: 'Loading failed!' }} />
      </IntlProvider>,
    )
    expect(queryByText(/Something went wrong/)).toBeInTheDocument()
  })

  it('should render the repositories if loading was successful', () => {
    const initialState = {
      [REDUCER_KEY_APP]: {
        currentUser: 'mxstbr',
        error: false,
        isFetching: false,
        userData: {
          repos: [],
        },
      },
    }
    const store = configureStore(initialState, history)
    const repos = [
      {
        owner: {
          login: 'mxstbr',
        },
        html_url: 'https://github.com/react-boilerplate/react-boilerplate',
        name: 'react-boilerplate',
        open_issues_count: 20,
        full_name: 'react-boilerplate/react-boilerplate',
      },
    ] as Repo[]
    const { container } = render(
      <Provider store={store}>
        <IntlProvider locale={DEFAULT_LOCALE}>
          <ReposList repos={repos} error={false} isFetching={false} />
        </IntlProvider>
      </Provider>,
    )

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should not render anything if nothing interesting is provided', () => {
    const { container, queryByText } = render(
      <IntlProvider locale={DEFAULT_LOCALE}>
        <ReposList repos={undefined} error={false} isFetching={false} />
      </IntlProvider>,
    )

    expect(container).not.toBeEmptyDOMElement()
    expect(queryByText(/Type user name in the input above and press Enter/)).toBeInTheDocument()
  })
})
