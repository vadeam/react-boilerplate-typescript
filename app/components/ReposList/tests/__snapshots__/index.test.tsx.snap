// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReposList /> should render the loading indicator when its loading 1`] = `
<div
  class="sc-gtsrHT cPrvMH"
>
  <ul
    class="sc-bdnxRM jzejmN"
  >
    <div
      class="sc-jSFjdj kxYqkO"
    >
      <div
        class="sc-eCApnc jzdQje"
      />
      <div
        class="sc-eCApnc gPsbou"
        rotate="30"
      />
      <div
        class="sc-eCApnc eVRtZn"
        rotate="60"
      />
      <div
        class="sc-eCApnc jBGUKg"
        rotate="90"
      />
      <div
        class="sc-eCApnc bcQaaS"
        rotate="120"
      />
      <div
        class="sc-eCApnc bgWsFF"
        rotate="150"
      />
      <div
        class="sc-eCApnc fxpZIo"
        rotate="180"
      />
      <div
        class="sc-eCApnc XYyCC"
        rotate="210"
      />
      <div
        class="sc-eCApnc lbTeUr"
        rotate="240"
      />
      <div
        class="sc-eCApnc dRyJFo"
        rotate="270"
      />
      <div
        class="sc-eCApnc gVOIOS"
        rotate="300"
      />
      <div
        class="sc-eCApnc iYcuXB"
        rotate="330"
      />
    </div>
  </ul>
</div>
`;

exports[`<ReposList /> should render the repositories if loading was successful 1`] = `
<div
  class="sc-gtsrHT cPrvMH"
>
  <ul
    class="sc-bdnxRM jzejmN"
  >
    <li
      class="sc-hKFxyN ibEoFF"
    >
      <div
        class="sc-dlnjwi fKaKZb"
      >
        <div
          class="sc-jrsJWt czIMqN"
        >
          <a
            class="sc-iCoGMd sc-pNWdM kLeFhJ iOqKJG"
            href="https://github.com/react-boilerplate/react-boilerplate"
            target="_blank"
          >
            react-boilerplate
          </a>
          <a
            class="sc-iCoGMd sc-fujyAs kLeFhJ kcJtpX"
            href="https://github.com/react-boilerplate/react-boilerplate/issues"
            target="_blank"
          >
            <svg
              class="sc-gKAaRy emMiVs"
              height="1em"
              width="0.875em"
            >
              <path
                d="M7 2.3c3.14 0 5.7 2.56 5.7 5.7S10.14 13.7 7 13.7 1.3 11.14 1.3 8s2.56-5.7 5.7-5.7m0-1.3C3.14 1 0 4.14 0 8s3.14 7 7 7 7-3.14 7-7S10.86 1 7 1z m1 3H6v5h2V4z m0 6H6v2h2V10z"
              />
            </svg>
            <span>
              20
            </span>
          </a>
        </div>
      </div>
    </li>
  </ul>
</div>
`;
